k2/csrc/tensor_ops.cu

```diff
@@ -507,4 +507,249 @@ Tensor SimpleRaggedIndexSelect1D(Tensor &src, Ragged<int32_t> &indexes) {
+    @param [in] y_stride0  Stride along axis 0 of of the `y` data
+    @param [in] stride1  Stride along axis 1 of the three arrays (this is expected
+                   to be identical, nonzero, and preferably -1 or 1.
+*/
+template <typename Real,
+          int ThreadsPerBlock>
+__global__ void DiscountedCumSumKernel(int N, int T,
+                                       const Real *x, int x_stride0,
+                                       const Real *gamma, int gamma_stride0,
+                                       Real *y, int y_stride0,
+                                       int stride1) {
+  int n_idx = blockIdx.y * gridDim.x + blockIdx.x;
+  if (n_idx >= N)
+    return;
+  x += x_stride0 * n_idx;
+  gamma += gamma_stride0 * n_idx;
+  y += y_stride0 * n_idx;
+
+  int thread_idx = threadIdx.x;
+  using Elem = DiscountedCumSumElement<Real>;
+
+  BlockPrefixCallbackOp<Real> prefix_callback;
+
+  typedef cub::BlockScan<Elem, ThreadsPerBlock> BlockScan;
+  // shared memory for BlockScan
+  __shared__ typename BlockScan::TempStorage temp_storage;
+
+  for (int base_t = 0; base_t < T; base_t += ThreadsPerBlock) {
+    Elem elem;
+
+    // Load x and gamma from memory.  These reads will be coalesced (which is
+    // the advantage of having each thread process one element at this stage;
+    // although we spend more time with raking reduction than we really
+    // need to).
+    if (base_t + thread_idx < T) {
+      elem.y = x[(base_t + thread_idx) * stride1];
+      elem.gamma = gamma[(base_t + thread_idx) * stride1];
+    }
+    CombineCumSumOp<Real> op;
+
+    // the last arg is a callback functor that provides us the aggregate of this block
+    // and which is expected to return the element that we want to add to
+    BlockScan(temp_storage).InclusiveScan(elem, elem, op, prefix_callback);
+
+    if ( base_t + thread_idx < T)
+      y[(base_t + thread_idx) * stride1] = elem.y;
+  }
+}
+
+
+template <typename Real, int ThreadsPerBlock>
+void DiscountedCumSumCudaImpl(cudaStream_t stream,
+                              int N, int T,
+                              const Real *x, int x_stride0,
+                              const Real *gamma, int gamma_stride0,
+                              Real *y, int y_stride0, int stride1) {
+
+  int32_t tot_grid_size = N;
+  int32_t x_grid_size = (tot_grid_size < (1 << 20) ?
+                         std::min<int32_t>(tot_grid_size, (1 << 10)) :
+                         32768),
+      y_grid_size = NumBlocks(tot_grid_size, x_grid_size);
+
+  dim3 grid_dim(x_grid_size, y_grid_size, 1),
+      block_dim(ThreadsPerBlock, 1, 1);
+  K2_CUDA_SAFE_CALL(DiscountedCumSumKernel<Real, ThreadsPerBlock>
+                    <<<grid_dim, block_dim, 0, stream>>>(N, T, x, x_stride0,
+                                                         gamma, gamma_stride0,
+                                                         y, y_stride0, stride1));
+}
+
+
+template <typename Real>
+void DiscountedCumSumCpuImpl(int N, int T,
```

Comment:
Should we make it `static` as it is supposed to be local in this translation unit?
(Also for the above CUDA kernel.)